File: src/pcms/pythonapi/test_field_copy.py
import py_pcms
import numpy as np

def test_copy(world, dim, order, num_components):
    """Test copying omega_h_field2 data"""
    nx = 100
    ny = 100 if dim > 1 else 0
    nz = 100 if dim > 2 else 0
    
    # Build mesh
    mesh = py_pcms.build_box(
        world, 
        py_pcms.OMEGA_H_SIMPLEX, 
        1.0, 1.0, 1.0, 
        nx, ny, nz, 
        False
    )
    
    # Create layout
    layout = py_pcms.CreateLagrangeLayout(
        mesh, 
        order, 
        num_components,
        py_pcms.CoordinateSystem.Cartesian
    )
    
    # Get number of data points
    ndata = layout.GetNumOwnedDofHolder() * num_components
    
    # Create sequential array of IDs
    ids = np.arange(ndata, dtype=np.float64)
    
    # Create original field and set data
    original = layout.CreateField()
    original.SetDOFHolderData(ids)
    
    # Create copied field and copy data
    copied = layout.CreateField()
    py_pcms.copy_field2(original, copied)
    
    # Get copied data
    copied_array = copied.GetDOFHolderData()
    
    # Verify the copy
    assert len(copied_array) == ndata, f"Expected {ndata} elements, got {len(copied_array)}"
    
    # Check that all values match
    differences = np.abs(ids - copied_array)
    num_matches = np.sum(differences < 1e-12)
    
    assert num_matches == ndata, f"Only {num_matches}/{ndata} elements matched"
    
    print(f"✓ Test passed: dim={dim}, order={order}, num_components={num_components}")

def main():
    """Run all test cases"""
    print("Testing copy omega_h_field2 data...")
    
    # Initialize Omega_h library
    lib = py_pcms.Library()
    world = lib.world()
    
    # Run test cases
    test_copy(world, 2, 1, 1)
    test_copy(world, 2, 2, 1)
    
    print("\nAll tests passed!")

if __name__ == "__main__":
    main()

File: test/test_field_copy.cpp
#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators.hpp>
#include <Omega_h_mesh.hpp>
#include <Omega_h_build.hpp>
#include <Omega_h_for.hpp>
#include <pcms/transfer_field2.h>
#include "pcms/adapter/meshfields/mesh_fields_adapter.h"
#include "pcms/adapter/meshfields/mesh_fields_adapter2.h"
#include "pcms/lagrange_field_factory.h"
#include "pcms/utility/assert.h"
#include <Kokkos_Core.hpp>

using pcms::Real;

void test_copy(Omega_h::CommPtr world, int dim, int order, int num_components)
{
  int nx = 100;
  int ny = dim > 1 ? 100 : 0;
  int nz = dim > 2 ? 100 : 0;

  auto mesh =
    Omega_h::build_box(world, OMEGA_H_SIMPLEX, 1, 1, 1, nx, ny, nz, false);
  auto factory = pcms::LagrangeFieldFactory::FromMesh(
    mesh, order, num_components, pcms::CoordinateSystem::Cartesian, "global",
    pcms::LagrangeFieldFactory::Backend::OmegaH);
  // FIXME this indicates that we are not exposing the right API from the fields to be able
  // to allocate buffers for getting/setting the field data.
  auto layout = factory.GetLayout();
  int ndata = layout->GetNumOwnedDofHolder() * num_components;
  Omega_h::HostWrite<Real> ids(ndata);
  Kokkos::parallel_for(
    Kokkos::RangePolicy<pcms::HostMemorySpace::execution_space>(0, ndata),
    [=](int i) { ids[i] = i; });

  auto original = factory.CreateFieldReal();
  original->SetDOFHolderData(pcms::make_const_array_view(ids));

  auto copied = factory.CreateFieldReal();
  pcms::copy_field2(*original, *copied);
  auto copied_array = copied->GetDOFHolderData();

  REQUIRE(copied_array.size() == ndata);
  int sum = 0;
  Kokkos::parallel_reduce(
    Kokkos::RangePolicy<pcms::HostMemorySpace::execution_space>(0, ndata),
    [=](int i, int& local_sum) {
      local_sum += std::abs(ids[i] - copied_array[i]) < 1e-12;
    },
    sum);
  REQUIRE(sum == ndata);
}

TEST_CASE("copy omega_h_field2 data")
{
  auto lib = Omega_h::Library{};
  test_copy(lib.world(), 2, 0, 1);
  test_copy(lib.world(), 2, 1, 1);
  auto mesh =
    Omega_h::build_box(lib.world(), OMEGA_H_SIMPLEX, 1, 1, 1, 100, 100, 0, false);
  REQUIRE_THROWS_AS(
    pcms::LagrangeFieldFactory::FromMesh(
      mesh,
      2, 1, pcms::CoordinateSystem::Cartesian, "global",
      pcms::LagrangeFieldFactory::Backend::OmegaH),
    pcms::pcms_error);
}
